### ISSUE TITLE
Adding FAUST and/or Pure Data?

### ISSUE THREAD

**Issue by frink** [christophhart/HISE#224]
Wondering if FAUST might be a good companion to HISE when extended effects are need for instruments... The FAUST interpreter could be embedded into HISE fairly easily for live editing and then used to generate C++ at compile time. This would allow for much more complex effects development without need to delve into C++ DSP.  The [stdfaust.lib](https://faustlibraries.grame.fr/) already have some advanced modeling for things like vocal tract, strings, bows, percussion, woodwinds, brass, and bells. Not to mention the colossal amount of open source FAUST that seems to be cropping up all the time. (e.g. [Guitarix](https://sourceforge.net/p/guitarix/git/ci/master/tree/trunk/src/faust/), [SEAM](https://github.com/s-e-a-m/faust-libraries), [Pultec EQ](https://github.com/droosenb/faust-wdf-library/blob/main/examples/03-pultec-eqp1a/pultec.dsp), [Yamaha YC-20](https://github.com/sampov2/foo-yc20), [Echoplex Ep-3](https://github.com/Rickr922/echoplex-sim), [Indian Tambura](https://github.com/olilarkin/Tambura), [String Machine](https://github.com/jpcima/string-machine), [Mass Interaction Models](https://github.com/mi-creative/MIMS), [Weather](https://github.com/myklemykle/weather_organ))

Likewise, PureData can be cross compiled to C++ as well via [HVCC](https://github.com/enzienaudio/hvcc) with a much longer history of open code available for perusal and reuse. It can also be embedded and interpreted for rapid development. FAUST and PD have very different paradigms allowing for very different developer experiences. Between these two I believe that you could Allow almost endless effects possibilities without the need for a deep understanding of real-time programming. _(NOTE: While CSound should also be a contender I didn't find any way to compile it to C++ - the clossest is [Cabbage](https://cabbageaudio.com/) which just embed it...)_

So while this is kind of a shot in the dark, (1 of 2 I'm proposing) my goal here is not merely to expand features but also to expand user-base. With this approach you're likely to add power-users who have been experimenting with sound-design for years. (This will likel give HISE a better name because they know what they are doing and just need to learn a bit of new syntax...) This has the potential of largely increasing the usage of HISE making development and testing much easier by spreading the load across many more parties.

As I see it there are two choices:

1. Support this in the tool itself.
2. Provide documentation for doing this externally.

I think it's likely to be a hybrid approach to allow for greater flexibility all the way around.

FAUST seems like much lower hanging fruit than PD because it is just a txt file and HISE is already a pretty powerful IDE. The only thing required would be a basic bridge to show the effects interfaces in HISE and a wrappers to get sound in and out of the FAUST code. If you really wanted to get exotic you could include an option to interpret FAUST code allowing for patches to have different DSP code and for the user to change aspects of the code without a need to recompile.

Thought it was worth getting a discussion going in the dev community rather than on the Forum. But feel free to edit this or move this to the Forum if that's a better place for discussion...

Cheers! ~M.Frink

**Comment by davidhealey** [comment]
hvcc is GNU GPL so not compatible with HISE commercial license. Pure Data is BSD, I'm not sure how the attribution clause would be enforced for derivative proprietary works.

I would like to see Faust integrated with HISE as I would like to use some of the physical modelling synthesis functionality. It has been discussed a couple of times on the HISE forum already:
https://forum.hise.audio/topic/2577/faust2hise
https://forum.hise.audio/topic/3226/faust-dsp-integration

**Comment by ustk** [comment]
Agree with both. I'd like to delve into Wave Digital Filters as well, I know we can find libraries for it in FAUST (or elsewhere too) but I didn't have the time to search for license compatibility yet for commercial projects.

**Comment by frink** [comment]
FAUST generated code is considered your own license so compiling FAUST generated DSP code should not be a problem. Incorporating the interpreter. The interpreter libfaust is GPL (I believe with a linking exception) and should be fine within HISE IDE but I'm not sure whether it would work within a plugin itself without license contamination. (need to ) There is already a [FAUST JUCE](https://github.com/olilarkin/juce_faustllvm) project underway which is GPL 3.0 could be referenced at the very least if not borrowed from directly...

> hvcc is GNU GPL so not compatible with HISE commercial license. Pure Data is BSD, I'm not sure how the attribution clause would be enforced for derivative proprietary works.

AFAIK hvcc and FAUST (both GPL 3.0 like HISE) treat generated code separate from the original files. (i.e. your code = your license) But we should probably dig deeper on that to be sure. FAUST stdfaust.lib has its [own license](https://github.com/grame-cncm/faustlibraries/blob/master/licenses/stk-4.3.0.md) which is similar to ISC and PD is BSD 3-clause as you have mentioned already. Either way we'll need a new bit of attribution in HISE alongside the acknowledgement of JUCE. Shouldn't be a big deal.

As an aside, it's probably worth it to include all the licenses (HISE + JUCE + FAUST + stdfaust.lib + PD) in the generated plugin anyway to make compliance easier on the end users. Bottom line, licenses should work so long as HISE doesn't become closed source. Should not contaminate generated plugins with GPL code other than what JUCE already does. 

Ugh! It seems like it takes a law degree to develop software anymore...

**Comment by dustbro** [comment]
FAUST + HISE would be a dream for me
+1000

**Comment by frink** [comment]
I wish I had time to do more than comment. Right now my dev time is overfull.

**Comment by sletz** [comment]
Faust developer here.  You guys can first read:
- the Faut [Architecture Files](https://faustdoc.grame.fr/manual/architectures/#developing-a-new-architecture-file) documentation
- and more specifically [Adapting the Generated DSP]( https://faustdoc.grame.fr/manual/architectures/#adapting-the-generated-dsp) 
- and [Developing New UI Architectures](https://faustdoc.grame.fr/manual/architectures/#developing-new-ui-architectures)

**Comment by frink** [comment]
Thanks @sletz! Would there be any interest in the FAUST community with helping to integrate into HISE? Chris is busy with so much stuff here I'm sure this is low priority.

**Comment by sletz** [comment]
Same here for now ))-;

**Comment by sletz** [comment]
GRAME has been [selected as a mentor organization](https://summerofcode.withgoogle.com/programs/2022/organizations/grame) for the Faust project.
And HISE integration if one of the [proposed projects](https://github.com/grame-cncm/faustideas/blob/master/README.md), hoping somewhere can be interested there?

**Comment by christoph-hart** [comment]
@sletz Nice to see HISE being included in the proposed projects. I'm happy to help out with the HISE integration. I think there are three levels of integration that might be implemented sequentielly:

1. Faust scriptnode architecture file. A exporter that spits out C++ from the Faust file that follows the API for scriptnode nodes (I'd suggest that all efforts of including Faust are being done on the scriptnode level as this is the way going forward for all DSP related things). The C++ code will then be compiled into a dll and will be loaded into scriptnode just like the C++ created by the scriptnode C++ compiler. For ultrafast results the existing system of loading the dynamic library with the compiled nodes of a project can be hacked.
2. Dynamic Faust code interpreter using LLVM. A node inside the scriptnode system (just like the other JIT nodes that use the SNEX compiler) that use the same conversion code from 1. but allow dynamic recompilation from a text file containing the Faust code.
3. IDE integration. HISE has a pretty extensive code editor system which already supports many languages (Javascript, C++, GLSL, XML) so a Faust syntax highlighter (based on the `JUCE::Tokeniser` class as well as the file management to work within HISE projects along with some other eye candy (eg. SVG signal graph display) might make the prototyping even more enjoyable.

I'm a bit reluctant about Faust integration myself at the moment because I spent so much time writing the SNEX JIT compiler which basically offers similar benefits (prototyping DSP with a live coding environment), however it lacks real world applications and a community that knows the language (it's a simple version of C++ but that needs to be learned too). So if there enough people who prefer writing DSP in Faust I think it's worthwhile to pursue it.

**Comment by sletz** [comment]
Thanks @christoph-hart for the detailed plan. Would you be OK to be GSoC co-mentor on the integration?

**Comment by christoph-hart** [comment]
It depends on the workload and schedule but yes that's definitely something I would enjoy to do.

**Comment by sletz** [comment]
The coding is between june and september, and mentoring means regular meeting with the contributor. I've added you name in the project, and we'll see how it goes.

**Comment by romsom** [comment]
Hi there, this is etXzat from the faust slack!
I'm interested in working on this as my project. Right now I'm still very busy with work and life and all that, but I'll try to squeeze in the time to have a look around and familiarize myself a bit with the code base. :)

**Comment by frink** [comment]
@christoph-hart -  I've done mentoring before in the past on GSoC. Currently swamped so I'm not a good choice now. But I can tell you that the impact on your schedule as a mentor is usually 0.5-1.5hrs/wk - not really a huge time commitment.

**Comment by romsom** [comment]
I'm reading documentation for ScriptNodes and libfaust and doing some code experiments at the moment.
As far as I can tell already I'd say the JIT approach is even more straight-forward to implement because all the heavy lifting is happening in Faust's DSPFactory.
The libfaust docs show the basics of it: https://faustdoc.grame.fr/manual/embedding/
The resulting dsp object can be connected to a "UI" that can manipulate its exported parameters and process audio buffers after it has been told the samplerate.
The wrapper ScriptNode would provide a fake "UI" to make the parameters available to HISE's scripting enginge.

I have a small prototype application running already that can process a static buffer of floats with jitted code.

@christoph-hart I'm gravitating towards skipping the DLL part in favor of using libfaust from the beginning. Are you fine with that or is there a benefit of part 1 of your integration plan with the dynamic library that I may be missing?

If I understand the docs correctly (@sletz: feel free to correct me if not), it is even possible to save and load the resulting code and we could bundle the compiled code to use that mechanism instead for a static approach.
I'll also prototype that and have a look at the resulting binary file format.
@frink That would probably correspond to part 3 of your proposed integration plan, i.e., provide a way to statically compile Faust code?

Also if I understand you correctly, in part 2 you suggest creating an Architecture File for Faust and a faust2hise script?
Can HISE export a project from the commandline possibly without spawning a gui? That would be a necessity here, I guess.

I'm writing my proposal at the moment and hope to have a first complete draft later today. I don't know if @sletz has already included you in the GSoC communication channels so you'll receive it as well when I upload it.
If not and if you're interested, I can send you a link to the drafts.
Final submission will be on Sunday, so there is not a lot of time left unfortunately, but maybe you still have time for some feedback as well.
I don't want to rob you of your Easter holidays, though :)

**Comment by sletz** [comment]
> it is even possible to save and load the resulting code and we could bundle the compiled code to use that mechanism instead for a static approach.

Yes using [save/restore kind of API](https://faustdoc.grame.fr/manual/embedding/#savingrestoring-the-factory_1)

@christoph-hart  you can possibly connect to the Faust Slack channel if needed: https://faust.grame.fr/community/help/

**Comment by sletz** [comment]
@frink Are you willing to be a GSoC for this project ? If yes I can add you [here](https://github.com/grame-cncm/faustideas/blob/master/README.md#integration-in-hise) ?

**Comment by sletz** [comment]
If @frink and @christoph-hart want to be mentors, I need to invite you on the GSoG web platform. Can you send me your emails? Thanks.

**Comment by christoph-hart** [comment]
> Are you fine with that or is there a benefit of part 1 of your integration plan with the dynamic library that I may be missing?

No that could also work (as I suggested a bit imprecisely with the phrase "Dynamic Faust code interpreter using LLVM"). The only thing that is important is that the compiled end product has no depencies on JIT compilation anymore as this would introduce a heavy overhead, so the creation of a Faust architecture file that will create C++ code is obligatory for the final step anyways.

> Can HISE export a project from the commandline possibly without spawning a gui?

Yes HISE can be used as command line tool to export projects (mainly for CI integration). However I think the preferred workflow would be:

1. Create a project in HISE
2. Add the Faust node in scriptnode, edit the Faust code and use the JIT compilation for instant prototyping.
3. Before (or during) the export, run a faust architecture file that converts the faust code to a "native" C++ node in scriptnode and add it to the project's C++ source files.
4. When you export the plugin in HISE, it will use the converted C++ files and embed it into the final binary without any dependencies on Faust JIT.

So inside HISE, Faust will always be JIT compiled and in the compiled projects it will be C++ code.

@romsom Yes please send me a link to the drafts. You can use the official `info at hise dot audio` email.

**Comment by romsom** [comment]
@christoph-hart Thanks for the clarification!
You should have just received an email.

After a discussion with @sletz yesterday, my understanding of our goals has improved quite a bit and that is also reflected in my current proposal.
To summarize the main goals are:
1. Implement dynamic/jit compilation (using a hard coded script until the editor is ready)
2. Integrate Faust into HISE's IDE
3. Implement static compilation (using transpilation to C++ and alternatively by using the save/restore API)

I will update the motivation for static compilation in my draft to include your arguments about prototyping.
The option to export a plugin with dynamic compilation I would leave in place for plugins where the code should be modifiable if you don't mind, though. :)

**Comment by romsom** [comment]
@christoph-hart Your email seems to have bounced, I'll try from another address.
Also the proposal has just been updated.
I'd be great if you could have a look at the detailed schedule, especially part 2, which is about the IDE integration.
Do you think the amount of time for that is realistic?

Thanks!

**Comment by christoph-hart** [comment]
Ah that's so annoying, I need to change the website provider :)

However I can read the proposal in the Google Summer of Code page now that @sletz added me as mentor.

The timeline sounds reasonable. The IDE integration might be straightforward as I've recently redesigned most IDE capabilities to work with both SNEX (C++) as well as Javascript, so adding a third Faust option can benefit from the new modularity. The autocomplete system can also register different providers which can be either extracted from the source code using doxygen or read any other XML file to load the Faust autocomplete data.

But let me know when you start digging into the codebase, this part is a bit underdocumented and "overabstracted" - I didn't expect anybody else to actually use / modify this section...

One thing that is really useful in HISE (both for SNEX and Javascript) is the live variable watch system that allows you to inspect variables in real time. I'm not sure whether the Faust JIT system allows live value inspection, but this would yield a significant advantage over standard solutions like VS Code et al (which is the competing workflow).

![live_watch](https://user-images.githubusercontent.com/6277426/163687054-494eff12-4806-4169-93bc-03083e6f6ebf.gif)

**Comment by sletz** [comment]
Thats is great ! 
We don't currently have ways to access the internal state of the DSP in JIT model. The way we usually do is to add kind of "probe" additional control outputs and visualise them.

**Comment by frink** [comment]
@sletz - Sadly I'm a little out of pocket with everything going right now. I'm in the middle of a physical studio build and three separate business endeavors. (Ugh. Don't ever try to do 2 startups at once!!!! - 3 is insanity!!!!) I'm sorry to say that I don't have time to do more than read comments once every week or two... ☹️

> To summarize the main goals are:
>
> - Implement dynamic/jit compilation (using a hard coded script until the editor is ready)
> - Integrate Faust into HISE's IDE
> - Implement static compilation (using transpilation to C++ and alternatively by using the save/restore API)

@romsom - Love it!

One other small feature that I'd like to put into this is the idea of using your own library instead of `stdfaust.lib`. The easiest way would be to load FAUST without any lib and `import("stdfaust.lib");` or any other to get your FAUSST workflow. Another thing to consider is where the include path is... (e.g. I might want to include some FAUST that is std to all my projects `stdfrink.lib` and other FAUST that is specific to my project...) Thinking through these issues now will encourage good project organization which will lead to better results in the end.

Glad to see this coming together!

**Comment by frink** [comment]
@romsom Just curious where things are at. Catching up on GH commentary and was surprised not to see anything here in a few months. Love to heave what's new.

**Comment by romsom** [comment]
@frink Looks pretty good actually. I'm wrapping up loose ends at the moment, so expect proper news in October!
If you are daunting and/or impatient you can check out my [dev branch](https://github.com/romsom/HISE/tree/etxzat-develop) here, which should probably work out of the box on linux, but only with the latest libfaust (master-dev) and with the interpreter backend enabled on windows. There is no documentation yet, but that'll arrive on [my blog](https://resonant-bytes.de/blog) in about two weeks.

**Comment by frink** [comment]
This is beautiful. Wondering about how the merge to master works in this project. Just realized there is a HUGE discrepancy...

![image](https://user-images.githubusercontent.com/18235/197293138-1e4cc77a-a9a7-491e-b91e-eaf428074d49.png)

I'm curious - @christoph-hart How does develop get merged to master?

**Comment by romsom** [comment]
Yup, but that's an issue for another day or week.
HISE has practically been living on the `develop` branch for quite some time now.

**Comment by sletz** [comment]
> Not to mention the colossal amount of open source FAUST that seems to be cropping up all the time. 
The more interesting projects are regularly added on the [Powered by Faust] (https://faust.grame.fr/community/made-with-faust/) page.

**Comment by christoph-hart** [comment]
> Wondering about how the merge to master works in this project

Yeah there's absolutely no reason to have a master branch this old, but everybody got used to pulling develop.

Anyways, I've merged the two now so HISE doesn't look like a abandoned project when you look at the master branch...

**Comment by dromer** [comment]
> AFAIK hvcc and FAUST (both GPL 3.0 like HISE) treat generated code separate from the original files. (i.e. your code = your license) But we should probably dig deeper on that to be sure.

Correct. Only the hvcc compiler tools are GPL 3.0 but the static c/c++ libraries are MIT and the transpiled code can be licensed to whatever the user wants.

Btw you'd probably want to use my branch of hvcc that contains a lot of updates and runs on modern systems: https://github.com/Wasted-Audio/hvcc
